src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs

```diff
@@ -40,52 +46,134 @@ internal SslStreamInternal(SslState sslState)
                 }
 
                 FreeReadBuffer();
             }
-            if (_PinnableOutputBuffer != null)
+            if (_pinnableOutputBuffer != null)
             {
                 if (PinnableBufferCacheEventSource.Log.IsEnabled())
                 {
-                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_PinnableOutputBuffer));
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_pinnableOutputBuffer));
                 }
 
-                s_PinnableWriteBufferCache.FreeBuffer(_PinnableOutputBuffer);
+                s_PinnableWriteBufferCache.FreeBuffer(_pinnableOutputBuffer);
             }
         }
 
         internal int Read(byte[] buffer, int offset, int count)
         {
-            return ProcessRead(buffer, offset, count);
+            return ProcessRead(buffer, offset, count, null);
         }
 
         internal void Write(byte[] buffer, int offset, int count)
         {
-            ProcessWrite(buffer, offset, count);
+            ProcessWrite(buffer, offset, count, null);
+        }
+
+        internal IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState)
+        {
+            BufferAsyncResult bufferResult = new BufferAsyncResult(this, buffer, offset, count, asyncState, asyncCallback);
+            AsyncProtocolRequest asyncRequest = new AsyncProtocolRequest(bufferResult);
+            ProcessRead(buffer, offset, count, asyncRequest);
+            return bufferResult;
+        }
+
+        internal int EndRead(IAsyncResult asyncResult)
+        {
+            if (asyncResult == null)
+            {
+                throw new ArgumentNullException("asyncResult");
+            }
+
+            BufferAsyncResult bufferResult = asyncResult as BufferAsyncResult;
+            if (bufferResult == null)
+            {
+                throw new ArgumentException(SR.Format(SR.net_io_async_result, asyncResult.GetType().FullName), "asyncResult");
+            }
+
+            if (Interlocked.Exchange(ref _nestedRead, 0) == 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndRead"));
+            }
+
+            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
+            bufferResult.InternalWaitForCompletion();
+
+            if (bufferResult.Result is Exception)
+            {
+                if (bufferResult.Result is IOException)
+                {
+                    throw (Exception)bufferResult.Result;
```

Comment:
I agree. I'm keeping the code as-is since we want to make parallel progress here and eventually in .Net Desktop. It's much easier if we keep the file-history.

Tracking the proposal with #5606.